pkg/backends/calico/client.go

```diff
@@ -811,6 +852,223 @@ func (c *client) onUpdates(updates []api.Update) {
+			c.globalListenPort = int(v3res.Spec.ListenPort)
+		}
+		c.updateCache(api.UpdateTypeKVUpdated, getKVPair(listenPortKey, strconv.Itoa(int(v3res.Spec.ListenPort))))
+	} else {
+		switch k := key.(type) {
+		case model.NodeBGPConfigKey:
+			delete(c.nodeListenPorts, getNodeName(k.Nodename))
+		case model.GlobalBGPConfigKey:
+			c.globalListenPort = 0
+		}
+		c.updateCache(api.UpdateTypeKVDeleted, getKVPair(listenPortKey))
+	}
+}
+
+func (c *client) getASNumberKVPair(v3res *apiv3.BGPConfiguration, key interface{}, updatePeersV1 *bool, updateReasons *[]string) {
+	asNumberKey := getBGPConfigKey("as_num", key)
+	if v3res != nil && v3res.Spec.ASNumber != nil {
+		*updateReasons = append(*updateReasons, "AS number updated.")
+		c.updateCache(api.UpdateTypeKVUpdated, getKVPair(asNumberKey, v3res.Spec.ASNumber.String()))
+	} else {
+		*updateReasons = append(*updateReasons, "AS number deleted.")
+		c.updateCache(api.UpdateTypeKVDeleted, getKVPair(asNumberKey))
+	}
+	*updatePeersV1 = true
+}
+
+func (c *client) getServiceExternalIPsKVPair(v3res *apiv3.BGPConfiguration, key interface{}, svcAdvertisement *bool) {
+	scvExternalIPKey := getBGPConfigKey("svc_external_ips", key)
+
+	if v3res != nil && v3res.Spec.ServiceExternalIPs != nil && len(v3res.Spec.ServiceExternalIPs) != 0 {
+		// We wrap each Service external IP in a ServiceExternalIPBlock struct to
+		// achieve the desired API structure, unpack that.
+		ipCidrs := make([]string, len(v3res.Spec.ServiceExternalIPs))
+		for i, ipBlock := range v3res.Spec.ServiceExternalIPs {
+			ipCidrs[i] = ipBlock.CIDR
+		}
+		c.updateCache(api.UpdateTypeKVUpdated, getKVPair(scvExternalIPKey, strings.Join(ipCidrs, ",")))
+	} else {
+		c.updateCache(api.UpdateTypeKVDeleted, getKVPair(scvExternalIPKey))
+	}
+	*svcAdvertisement = true
+}
+
+func (c *client) getServiceClusterIPsKVPair(v3res *apiv3.BGPConfiguration, key interface{}, svcAdvertisement *bool) {
+	svcInternalIPKey := getBGPConfigKey("svc_cluster_ips", key)
+
+	if v3res != nil && v3res.Spec.ServiceClusterIPs != nil && len(v3res.Spec.ServiceClusterIPs) != 0 {
+		if len(os.Getenv(envAdvertiseClusterIPs)) != 0 {
+			// ClusterIPs are configurable through an environment variable. If specified,
+			// that variable takes precedence over datastore config, so we should ignore the update.
+			// Setting Spec.ServiceClusterIPs to nil, so we keep using the cache value set during startup.
+			log.Info("Ignoring serviceClusterIPs update due to environment variable %s", envAdvertiseClusterIPs)
+		} else {
+			// We wrap each Service Cluster IP in a ServiceClusterIPBlock to
+			// achieve the desired API structure. This unpacks that.
+			ipCidrs := make([]string, len(v3res.Spec.ServiceClusterIPs))
+			for i, ipBlock := range v3res.Spec.ServiceClusterIPs {
+				ipCidrs[i] = ipBlock.CIDR
+			}
+			*svcAdvertisement = true
+			c.updateCache(api.UpdateTypeKVUpdated, getKVPair(svcInternalIPKey, strings.Join(ipCidrs, ",")))
+		}
+	} else {
+		*svcAdvertisement = true
```

Comment:
I think this line is wrong, in the case where cluster IPs are configured in the environment; it will cause the environment configuration to be overwritten with nothing.

Believe the fix is to promote the `len(os.Getenv(envAdvertiseClusterIPs)) != 0` check so that it applies to all of the code in this function.